File: iosApp/iosApp/ui/storiesList/StoriesListViewModel.swift
import Foundation
import shared

extension StoriesListScreen {
    @MainActor class StoriesListViewModel : ObservableObject, ViewLifeCycleObserver {
        
        private var commonViewModel: StoriesListCommonViewModel?
        
        @Published var storiesLoadStatus: LoadStatus<Reference<[History]>>? = nil
        @Published var newHistory: History? = nil
        @Published var isLogged: Bool? = nil
        
        func deleteHistory(historyId: String) {
            commonViewModel?.deleteHistory(historyId: historyId)
        }

        func createBasicHistory(title: String, text: String) {
            commonViewModel?.createBasicHistory(title: title, text: text)
        }

        func onNewHistoryConsumed() {
            commonViewModel?.onNewHistoryConsumed()
        }

        func startObserving() {
            let commonViewModel = StoriesListCommonViewModel()
            self.commonViewModel = commonViewModel
            commonViewModel.storiesLoadStatus.subscribe(scope: commonViewModel.viewModelScope) { stories in
                if let stories = stories {
                    self.storiesLoadStatus = stories.mapData { stories in
                        return Reference(value: stories as? [History] ?? [])
                    } as? LoadStatus<Reference<[History]>>
                }
            }
            commonViewModel.newHistory.subscribe(scope: commonViewModel.viewModelScope) { newHistory in
                self.newHistory = newHistory
            }
            commonViewModel.isLogged.subscribe(scope: commonViewModel.viewModelScope) { isLogged in
                self.isLogged = isLogged?.boolValue
            }
        }
        
        func stopObserving() {
            commonViewModel?.dispose()
        }
    }
}

File: androidApp/src/main/java/com/example/stories/android/ui/storiesList/StoriesListViewModel.kt
package com.example.stories.android.ui.storiesList

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.stories.viewModel.StoriesListCommonViewModel

class StoriesListViewModel : ViewModel() {

    private val commonViewModel = StoriesListCommonViewModel(viewModelScope)

    val storiesLoadStatus get() = commonViewModel.storiesLoadStatus
    val newHistory get() = commonViewModel.newHistory
    val isLogged get() = commonViewModel.isLogged

    fun deleteHistory(historyId: String) = commonViewModel.deleteHistory(historyId)
    fun createBasicHistory(title: String, text: String) = commonViewModel.createBasicHistory(title, text)
    fun onNewHistoryConsumed() = commonViewModel.onNewHistoryConsumed()
}
